src/xrpld/app/misc/NetworkOPs.cpp

```diff
@@ -1441,16 +1540,28 @@ NetworkOPsImp::apply(std::unique_lock<std::mutex>& batchLock)
                     << "Transaction is now included in open ledger";
                 e.transaction->setStatus(INCLUDED);
 
+                // Pop as many "reasonable" transactions for this account as
+                // possible. "Reasonable" means they have sequential sequence
+                // numbers, or use tickets.
                 auto const& txCur = e.transaction->getSTransaction();
-                auto const txNext = m_ledgerMaster.popAcctTransaction(txCur);
-                if (txNext)
+
+                std::size_t count = 0;
+                for (auto txNext = m_ledgerMaster.popAcctTransaction(txCur);
+                     txNext && count < maxPoppedTransactions;
+                     txNext = m_ledgerMaster.popAcctTransaction(txCur), ++count)
                 {
+                    if (!batchLock.owns_lock())
+                        batchLock.lock();
                     std::string reason;
                     auto const trans = sterilize(*txNext);
                     auto t = std::make_shared<Transaction>(trans, reason, app_);
+                    if (t->getApplying())
+                        break;
                     submit_held.emplace_back(t, false, false, FailHard::no);
                     t->setApplying();
                 }
+                if (batchLock.owns_lock())
+                    batchLock.unlock();
```

Comment:
I don't love that this code **needs** to do it, but I agree that's the best (most robust) way to deal with the problem. Approved.

Reply:
> I don't love that this code **needs** to do it, but I agree that's the best (most robust) way to deal with the problem. Approved.

Yeah, it's a compromise between not wanting to take the lock outside the loop, which would lock every time a transaction succeeds, not knowing if there are any transactions to pop, and not wanting to take the lock inside the loop, which would churn a lot for what should be a very quick loop.